### PR TITLE
Implement the instant finish of the battle

### LINKED ISSUE #5626: Add an option to complete battle instantly
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.

### Describe the problem requiring a solution

Sometimes players want to finish battle as soon as possible, but they don't want to run away or surrender. The only way to do this currently is to set maximum speed and initiate auto battle. Or continue playing manually.

### Describe the possible solution

I think we should implement at least hot key for now to jump to the battle outcome, as if player pressed AUTO.

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
close #5626

Currently, only the hotkey can be used (`Q` by default - yes, just like in HoTA). There is no graphical button for this.

https://user-images.githubusercontent.com/32623900/185990658-f3099b04-7db1-4e0b-af83-760fa9d26b35.mp4

Currently, this works both in single-player and in hotseat mode, because, since now both human players are able to use the instant battle and auto combat, I do not see a reason to disable the instant finish in hotseat (and yes, it is convenient for debugging as well).

Also this PR fixes potential issues with non-deterministic random generator for Genie attack, Chain Lighting troop resistance and Earth Quake damage.

### CODE REVIEW

File: src/fheroes2/battle/battle_interface.cpp [hunk 1]
```diff
@@ -2708,10 +2716,24 @@ void Battle::Interface::EventShowOptions()
...
 
+    a.emplace_back( CommandType::MSG_BATTLE_AUTO_SWITCH, b.GetCurrentOrArmyColor() );
+
+    humanturn_redraw = true;
+    humanturn_exit = true;
+}
+
+void Battle::Interface::EventAutoFinish( Actions & a )
+{
+    if ( Dialog::Message( "", _( "Are you sure you want to finish the battle in auto mode?" ), Font::BIG, Dialog::YES | Dialog::NO ) != Dialog::YES ) {
```
Comment (ihhub) [inline src/fheroes2/battle/battle_interface.cpp:2731]: Please use the newest version of UI Dialog:
```cpp
if ( fheroes2::showMessage( fheroes2::Text( "", {} ),
     fheroes2::Text( _( "Are you sure you want to finish the battle in auto mode?" ), fheroes2::FontType::normalWhite() ), Dialog::YES | Dialog::NO ) != Dialog::YES ) {
    return;
}
```

I know it's much longer but we might come up with some shorter solution in the future.

**Review by ihhub — COMMENTED** [review]
Hi @oleg-derevenetz . everything looks good except one place where we should use newer version of UI calls. Please take a look.

File: src/fheroes2/battle/battle_interface.cpp (hunk 1 above)
Comment (oleg-derevenetz) [reply src/fheroes2/battle/battle_interface.cpp]: Done in both places.

**Comment by ihhub** [comment]
@oleg-derevenetz , a bunch of thanks for implementing this feature!

**Pull request merged**